worker/src/sidechain_storage/storage.rs

```diff
@@ -0,0 +1,1062 @@
+/*
+	Copyright 2019 Supercomputing Systems AG
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+		http://www.apache.org/licenses/LICENSE-2.0
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+*/
+
+use super::{db::SidechainDB, Error, Result};
+use codec::{Decode, Encode};
+use log::*;
+use rocksdb::WriteBatch;
+use sp_core::H256;
+use std::{collections::HashMap, fmt::Debug, path::PathBuf};
+use substratee_worker_primitives::{
+	block::{BlockHash, BlockNumber},
+	traits::{Block as BlockT, SignedBlock as SignedBlockT},
+};
+/// key value of sidechain db of last block
+const LAST_BLOCK_KEY: &[u8] = b"last_sidechainblock";
+/// key value of the stored shards vector
+const STORED_SHARDS_KEY: &[u8] = b"stored_shards";
+
+/// ShardIdentifier type
+type ShardIdentifierFor<B> = <<B as SignedBlockT>::Block as BlockT>::ShardIdentifier;
+/// Helper struct, contains the blocknumber
+/// and blockhash of the last sidechain block
+#[derive(PartialEq, Eq, Clone, Copy, Encode, Decode, Debug, Default)]
+pub struct LastSidechainBlock {
+	/// hash of the last sidechain block
+	pub hash: H256,
+	/// block number of the last sidechain block
+	pub number: BlockNumber,
+}
+
+/// Struct used to insert newly produced sidechainblocks
+/// into the database
+pub struct SidechainStorage<SignedBlock: SignedBlockT> {
+	/// database
+	db: SidechainDB,
+	/// shards in database
+	shards: Vec<ShardIdentifierFor<SignedBlock>>,
+	/// map to last sidechain block of every shard
+	last_blocks: HashMap<ShardIdentifierFor<SignedBlock>, LastSidechainBlock>,
+}
+
+impl<SignedBlock: SignedBlockT> SidechainStorage<SignedBlock> {
+	/// loads the DB from the given paths and stores the listed shard
+	/// and their last blocks in memory for better performance
+	pub fn new(path: PathBuf) -> Result<SidechainStorage<SignedBlock>> {
+		// load db
+		let db = SidechainDB::open_default(path)?;
+		let mut storage = SidechainStorage { db, shards: vec![], last_blocks: HashMap::new() };
+		storage.shards = storage.load_shards_from_db()?;
+		// get last block of each shard
+		for shard in storage.shards.iter() {
+			if let Some(last_block) = storage.load_last_block_from_db(shard)? {
+				storage.last_blocks.insert(*shard, last_block);
+			} else {
+				// an empty shard sidechain storage should not exist. Consider deleting this shard from the shards list.
+				error!("Sidechain storage of shard {:?} is empty", shard);
+			}
+		}
+		Ok(storage)
+	}
+
+	/// gets all shards of currently loaded sidechain db
+	pub fn shards(&self) -> &Vec<ShardIdentifierFor<SignedBlock>> {
```

Comment:
It does seem to be something different. But you might be able to help me here:
https://github.com/integritee-network/worker/blob/sidechain-storage/worker/src/sidechain_storage/storage.rs#L226

if I do not clone here, it will complain about once borrowing as immutable and once as mutable (here: https://github.com/integritee-network/worker/blob/sidechain-storage/worker/src/sidechain_storage/storage.rs#L230)

So my solution: clone it once to get the sharsd. However, if I change the &Vec to &[] the compiler can not be statisfied at all. 

What would be your approach there?

Reply:
Ah you are right. I think I got it wrong. It is probably better to return a reference to the owned type instead of the borrowed type. Will keep that in mind. Sorry for the circumstances

Reply:
Not a problem. This is a topic I need to do some further reading anyway.